### PR TITLE
threaded alloc

### PR DESCRIPTION
**Pull request by Araq**
- WIP
- Don't use locks for small allocations but allow a foreign thread to dealloc an object
- Big allocations take thread-specific locks

### CODE REVIEW

**Comment by Araq** [comment]
Ping @mratsim. Your feedback would be highly appreciated.

**Comment by Araq** [comment]
https://github.com/nim-lang/Nim/pull/20492/files#diff-a0e0c2de7fc0517f631eb8f43a521bc49d7e33532c7b8362c5d8f9699aeec1ddR728-R733 is supposed to implement a lockfree list operation. But it seems to race with https://github.com/nim-lang/Nim/pull/20492/files#diff-a0e0c2de7fc0517f631eb8f43a521bc49d7e33532c7b8362c5d8f9699aeec1ddR751 which is quite impossible as that iterates over a list that has been stolen successfully via https://github.com/nim-lang/Nim/pull/20492/files#diff-a0e0c2de7fc0517f631eb8f43a521bc49d7e33532c7b8362c5d8f9699aeec1ddR793

File: lib/system/alloc.nim [hunk 1]
```diff
@@ -44,6 +44,29 @@ type
...
+    next: PLLChunk           # next low-level chunk; only needed for dealloc
+
+  PAvlNode = ptr AvlNode
+  AvlNode = object
+    link: array[0..1, PAvlNode] # Left (0) and right (1) links
+    key, upperBound: int
+    level: int
+
+const
+  RegionHasLock = hasThreadSupport and defined(gcDestructors)
```
Comment (mratsim) [inline lib/system/alloc.nim]: Don't refc and markandsweep also lock?

File: lib/system/alloc.nim [hunk 2]
```diff
@@ -117,17 +120,27 @@ type
     llmem: PLLChunk
     currMem, maxMem, freeMem, occ: int # memory sizes (allocated from OS)
     lastSize: int # needed for the case that OS gives us pages linearly
+    when RegionHasLock:
+      lock: SysLock
     chunkStarts: IntSet
     when not defined(gcDestructors):
       root, deleted, last, freeAvlNodes: PAvlNode
-    locked, blockChunkSizeIncrease: bool # if locked, we cannot free pages.
+    lockActive, locked, blockChunkSizeIncrease: bool # if locked, we cannot free pages.
```
Comment (mratsim) [inline lib/system/alloc.nim:131]: It's unclear the difference between `locked` and `lockActive`, especially because locked doesn't appear in this PR

Comment (mratsim) [reply lib/system/alloc.nim:131]: looking at the full files `locked` is set in the iterator `allObjects` but never checked in code?

File: lib/system/alloc.nim [hunk 3]
```diff
@@ -234,17 +247,17 @@ proc addChunkToMatrix(a: var MemRegion; b: PBigChunk) =
   setBit(fl, a.flBitmap)
 
 proc incCurrMem(a: var MemRegion, bytes: int) {.inline.} =
-  inc(a.currMem, bytes)
+  atomicInc(a.currMem, bytes)
```
Comment (mratsim) [inline lib/system/alloc.nim]: For review I think it's important to mention that `atomicInc` uses sequential consistency https://github.com/nim-lang/Nim/blob/b6f8b5f/lib/system/atomics.nim#L218-L226

File: lib/system/alloc.nim (hunk 2 above)
Comment (mratsim) [reply lib/system/alloc.nim:131]: Ah `lockActive` means lock is initialized

File: lib/system/alloc.nim [hunk 4]
```diff
@@ -61,12 +84,15 @@ type
...
     size: int            # if < PageSize it is a small chunk
+    owner: ptr MemRegion
 
   SmallChunk = object of BaseChunk
     next, prev: PSmallChunk  # chunks of the same size
     freeList: ptr FreeCell
     free: int            # how many bytes remain
     acc: int             # accumulator for small object allocation
+    when defined(gcDestructors):
+      sharedFreeList: ptr FreeCell # make no attempt at avoiding false sharing for now for this object field
```
Comment (mratsim) [inline lib/system/alloc.nim:95]: For documentation this should be an atomic

**Review by mratsim — COMMENTED** [review]
Just to make sure we agree on the intended threaded flow:

1. Thread 1 allocates a small chunk
2. The user application moves that memory on thread 2, it is reclaimed there
3. Because of the transfer of ownership, the memory is added to a "shared" free list. Though only thread 2 as a pointer left to it at this point.
4. If that chunk is picked for allocation, the sharedFreeList and ownership of it are transferred to freeList

**Comment by Araq** [comment]
> Just to make sure we agree on the intended threaded flow: ...

Correct.

File: lib/system/alloc.nim (hunk 3 above)
Comment (Araq) [reply lib/system/alloc.nim]: Good point and we only need `ATOMIC_RELAXED` here.

File: lib/system/alloc.nim (hunk 2 above)
Comment (Araq) [reply lib/system/alloc.nim:131]: Correct.

File: lib/system/alloc.nim (hunk 1 above)
Comment (Araq) [reply lib/system/alloc.nim]: No, the old GCs remain thread-local and so don't need a lock for the memory region.

**Comment by mratsim** [comment]
So is this a temporary adaptation until mimalloc is integrated more deeply into Nim?

File: lib/system/alloc.nim [hunk 5]
```diff
@@ -44,6 +44,29 @@ type
...
+    link: array[0..1, PAvlNode] # Left (0) and right (1) links
+    key, upperBound: int
+    level: int
+
+const
+  RegionHasLock = hasThreadSupport and defined(gcDestructors)
+
 type
   FreeCell {.final, pure.} = object
     next: ptr FreeCell  # next free cell in chunk (overlaid with refcount)
```
Comment (mratsim) [inline lib/system/alloc.nim:72]: As mentioned in the forum, this should be `Atomic[ptr FreeCell]` hence why ThreadSanitizer warns.

File: lib/system/alloc.nim [hunk 6]
```diff
@@ -691,6 +724,35 @@ else:
   template trackSize(x) = discard
   template untrackSize(x) = discard
 
+when defined(gcDestructors):
+  proc addToSharedFreeList(c: PSmallChunk; f: ptr FreeCell) {.inline.} =
+    # see also https://en.cppreference.com/w/cpp/atomic/atomic_compare_exchange
+    while true:
+      f.next = c.sharedFreeList.loada
```
Comment (mratsim) [inline lib/system/alloc.nim]: f.next should be an atomicStore

File: lib/system/alloc.nim [hunk 7]
```diff
@@ -691,6 +724,35 @@ else:
...
+    var x = 0
+    #var maxIters = 20 # make it time-bounded
+    while it != nil:
+      #if maxIters == 0:
+      #  let rest = it.next.loada
+      #  it.next = nil
+      #  addToSharedFreeList(c, rest)
+      #  break
+      inc x, size
+      it = it.next
```
Comment (mratsim) [inline lib/system/alloc.nim]: it.next should be an atomicLoad

**Review by mratsim — COMMENTED** [review]
Re ThreadSanitizer warning

**Comment by Araq** [comment]
> So is this a temporary adaptation until mimalloc is integrated more deeply into Nim?

It's an alternative to mimalloc which keeps TLSF's benefits and downsides. Benchmarks should eventually decide which one to make the default but TLSF has different fragmentation behavior so it doesn't hurt to have both around.

**Pull request merged**